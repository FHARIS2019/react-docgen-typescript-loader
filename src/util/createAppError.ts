interface AppError extends Error {
  innerError?: Error;
}

/**
 * Creates an error message using the provided error object or error message. It
 * wraps the message with the name of the plugin to make it easier to track down
 * issues when debugging build errors.
 *
 * @param error Error message or error object to wrap.
 */
function createAppError(error?: Error | string): Error {
  const errorMessage = error
    ? typeof error === "string" ? error : error.message
    : "An unexpected error has occurred.";

  const appError: AppError = new Error(
    `react-docgen-typescript-webpack-plugin: ${errorMessage}`,
  );

  appError.innerError = typeof error === "string" ? undefined : error;

  Error.captureStackTrace(appError, createAppError);

  return appError;
}

export default createAppError;
