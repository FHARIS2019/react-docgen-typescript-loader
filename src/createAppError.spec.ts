import createAppError from "./createAppError";

it("returns an error when no error object is provided", () => {
  expect(createAppErrorAsJson()).toMatchSnapshot();
});

it("wraps a provided error object", () => {
  expect(createAppErrorAsJson(new Error("Provided Error"))).toMatchSnapshot();
});

it("creates error from message", () => {
  expect(createAppErrorAsJson("Error Message")).toMatchSnapshot();
});

function createAppErrorAsJson(error?: string | Error) {
  const appError: Error & { innerError?: Error } = createAppError(error);

  return {
    message: appError.message,
    stack: appError.stack && appError.stack.split("\n")[0],
    innerError: appError.innerError && appError.innerError.message,
  };
}
