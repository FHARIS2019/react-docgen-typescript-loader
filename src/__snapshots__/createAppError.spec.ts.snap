// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates error from message 1`] = `
Object {
  "innerError": undefined,
  "message": "react-docgen-typescript-loader: Error Message",
  "stack": "Error: react-docgen-typescript-loader: Error Message",
}
`;

exports[`returns an error when no error object is provided 1`] = `
Object {
  "innerError": undefined,
  "message": "react-docgen-typescript-loader: An unexpected error has occurred.",
  "stack": "Error: react-docgen-typescript-loader: An unexpected error has occurred.",
}
`;

exports[`wraps a provided error object 1`] = `
Object {
  "innerError": "Provided Error",
  "message": "react-docgen-typescript-loader: Provided Error",
  "stack": "Error: react-docgen-typescript-loader: Provided Error",
}
`;
